File: c/binary/binary.c
#include "binary.h"


int convert(const char *p) {
    int result = 0;
    for (; *p; p++) {
        int digit = *p - '0';
        if (!(0 <= digit && digit < BASE))
            return INVALID;
        result = result * BASE + digit;
    }
    return result;
}

File: php/binary/Binary.php
<?php

declare(strict_types=1);

function parse_binary(string $binary): int
{
    return array_reduce(
        str_split($binary),
        function ($decimal, $digit) {
            switch ($digit) {
            case '0':
            case '1':
                return $decimal << 1 | $digit;
            default:
                throw new InvalidArgumentException('not a binary digit');
            };
        },
        0
    );
}
